v1.0

Anarmard opened the pull request:
- add booking and comments
- DB instead of memory

Проблема: даты в booking. Формат

Anarmard committed:
v2.0
- add booking and comments
- DB instead of memory

Anarmard committed:
v2.1
- add booking and comments
- DB instead of memory
- checkstyle

alexache commented on src/main/java/ru/practicum/shareit/booking/dto/BookingCreateRequestDto.java:
```diff
@@ -0,0 +1,33 @@
+
+    private Long id; // уникальный идентификатор бронирования
+
+    @NotBlank
```
@NotBlank не очень применимы для дат, лучше использовать @Future, @FutureOrPresent и прочие более релевантные

Anarmard replied:
спасибо, поправил

alexache commented on src/main/java/ru/practicum/shareit/booking/service/BookingServiceImpl.java:
```diff
@@ -0,0 +1,157 @@
+    public List<Booking> getAllBooking(Long userId, String state) {
+        List<Booking> result;
+        User booker = userRepository.findById(userId).orElseThrow(() -> new NotFoundException("Booking is not found"));
+        Timestamp currentMoment = Timestamp.valueOf(LocalDateTime.now());
```
Почему Timestamp?  Лучше не отходить нигде от LocalDateTime, с остальным Hibernate справится

Anarmard replied:
Спасибо за совет, исправил!

Anarmard committed:
v2.2
- LocalDate (Timestamp)
- additional checks

Anarmard committed:
v2.3
- добавил различные проверки booking

Anarmard committed:
v2.3
- добавил различные проверки booking

Anarmard committed:
v2.3
- добавил различные проверки booking

Anarmard committed:
v2.3
- добавил различные проверки booking

Anarmard committed:
v2.3
- добавил различные проверки booking

Anarmard committed:
v2.8
- поиск ошибки

Anarmard committed:
v2.9
- поиск ошибки

Anarmard committed:
v2.10
- исправил название столбцов в таблице ITEMS

Anarmard committed:
v2.11
- исправил название столбцов в Comment (чтобы совпадали с SQL запросом)

Anarmard commented:
Александр, привет! За эту неделю пока правил программу, узнал, что если проходят все тесты в Postman, тесты на гитхабе могут не проходить. Ошибка была в том, что в объектах Item и Comment я указал неверные названия столбцов... Не понимаю тогда, почему тесты в постмане проходились.

alexache commented on src/main/java/ru/practicum/shareit/booking/controller/BookingController.java:
```diff
@@ -0,0 +1,79 @@
+    @PostMapping
+    public BookingResponseDto saveBooking(@RequestHeader("X-Sharer-User-Id") Long userId,
+                               @Valid @RequestBody BookingCreateRequestDto bookingCreateRequestDto) {
+        userService.getUserById(userId);
```
(опционально) на мой взгляд, выходит слишком много логики для контроллера - столько коммуникации между вызовами-мапперы-другие сущности лучше решать в сервисе, а из сервиса вызывать репозитории соответствующих сущностей

Anarmard replied:
Задумка была следующая: 
1. Чтобы в контроллере происходил перевод из ДТО в модель. Затем модель отправляется в сервис. Ответ из сервиса также возвращается в виде модели. И уже в контроллере происходит перевод из модели в ДТО.

2. Также часто использую метод: **userService.getUserById(userId),** чтобы проверить на уровне контроллера, существует ли такой User.

3. Если в контроллере оставить только вызов метода из сервиса. А в сервисе уже устраивать все проверки и переводы из ДТО в модель и обратно, то получается, что контроллер будет пропускать несуществующих User.
Это нормально? Если это ок, то давай я лучше поправлю этот момент.

alexache replied:
3. Да, это ок. Обычно контроллеры валидируют данные, но не всегда с запросами к базе, это можно оставить на сервис

Anarmard replied:
Облегчил по максимуму все контроллеры, перевел всё в сервисы

alexache commented on src/main/java/ru/practicum/shareit/booking/dto/BookingCreateRequestDto.java:
```diff
@@ -0,0 +1,31 @@
+import javax.validation.constraints.FutureOrPresent;
+import java.time.LocalDateTime;
+
+@Valid
```
Обычно эту аннотацию ставят в контроллере непосредственно на аргумент

Anarmard replied:
Получается это здесь лишнее (в контроллере данную аннотацию использую). Убрал. Спасибо, Александр!

alexache commented on src/main/java/ru/practicum/shareit/booking/dto/BookingResponseDto.java:
```diff
@@ -1,21 +1,22 @@
- */
-@Data
-public class BookingDto {
+@Valid
```
А для ответа от контроллера применимость данной аннотации сомнительна

Anarmard replied:
Получается в ответных dto странно использовать аннотацию valid. Убрал. Спасибо, Александр!

alexache commented on src/main/java/ru/practicum/shareit/booking/repository/BookingRepository.java:
```diff
@@ -0,0 +1,63 @@
+
+    List<Booking> findAllByBookerOrderByStartDesc(User booker);
+
+    @Query("select b from Booking b " +
```
(опционально) Многие (если не все) запросы из кастомных @Query здесь можно бы реализовать через именование методов, как в случае с самым первым методом репозитория выше. Запрос будет немного сложнее, и есть пределы, когда это перестает быть целесообразно (длина метода превышает разумные границы)

Anarmard replied:
Если честно я начинал писать запросы через именование методов (запросные методы). Но на каких-то запросах вылетали ошибки, потратил время на их исправление, в итоге написал как подсказал наставник.

alexache commented on src/main/java/ru/practicum/shareit/booking/service/BookingServiceImpl.java:
```diff
@@ -0,0 +1,166 @@
+        LocalDateTime currentMoment = LocalDateTime.now();
+
+        // переводим из String в ENUM
+        BookingState stateBooking = Objects.isNull(state) ? BookingState.ALL : BookingState.of(state);
```
(опционально) это можно решать выставлением значений по умолчанию у аргументов контроллера

Anarmard replied:
Сделано! спасибо

alexache commented on src/main/java/ru/practicum/shareit/item/dto/CommentCreateRequestDto.java:
```diff
@@ -0,0 +1,27 @@
+
+    private Item item;
+
+    private User authorName;
```
Наличие неДТО полей в ДТО классе не очень хороший знак. Лучше разделять модели и ДТО так, чтобы они не смешивались

Anarmard replied:
Да, точно! Спасибо! исправил

alexache commented on src/main/java/ru/practicum/shareit/item/service/ItemService.java:
```diff
@@ -10,11 +13,15 @@ public interface ItemService {
     Item getItem(Long itemId);
 
-    List<Item> getItems(Long userId);
+    List<ItemBookingResponseDto> getItemsBooking(Long userId);
```
Тоже смешивание ДТО и моделей в  качестве возвращаемого типа в одном интерфесе, что выглядит немного настораживающе

Anarmard replied:
Задумка была следующая: 
чтобы в контроллере происходил перевод из ДТО в модель. Затем модель отправляется в сервис. Ответ из сервиса также возвращается в виде модели. И уже в контроллере происходит перевод из модели в ДТО.
Но в данных методах так сделать не получилось, т.к. поля lastbooking & nextbooking есть только ItemBookingResponseDto... В item таких полей нет.

alexache commented on src/main/java/ru/practicum/shareit/item/service/ItemServiceImpl.java:
```diff
@@ -2,54 +2,145 @@
+        // только владелец вещи может видеть последнее и будущее бронирование
+        if (Objects.equals(itemFromDB.getOwner().getId(), userId)) {
+            // нашли бронирования у данного item - последнее и будущее
+            List<Booking> lastListBooking = bookingRepository.findLastBooking(itemFromDB, BookingStatus.APPROVED, LocalDateTime.now());
```
Интересно, отчего метод возвращения последнего бронирования может возвращать список

Anarmard replied:
Я пробовал ограничить выдачу запроса 1 элементом (через limit 1), но не сработало.... Как я понял с форума: 
JPQL (то, что использую я вместо запросных методов) не поддерживает выдачу лимитированного ответа.

alexache commented on src/main/java/ru/practicum/shareit/item/service/ItemServiceImpl.java:
```diff
@@ -2,54 +2,145 @@
+        List<Item> itemListFromDB = itemRepository.findItemsByOwner(userFromDB);
+        List<ItemBookingResponseDto> itemBookingResponseDtoList = new ArrayList<>();
+        for (Item itemFromDB : itemListFromDB) {
+            itemBookingResponseDtoList.add(getItemBooking(itemFromDB.getId(), userId));
```
Запросы к базе в цикле под большим табу, потому что это выглядит с виду просто, а по факту очень непредсказуемый рост затрат по времени и ресурсам. Обычно это обходится отдельным запросом к базе со списком сущностей - получить последний бронирования для списка товаров и дальнейшей соответствующей обработки уже в джаве

Anarmard replied:
Понял. Переписал метод, вроде работает.

alexache commented on src/main/java/ru/practicum/shareit/item/service/ItemServiceImpl.java:
```diff
@@ -2,54 +2,145 @@
+    @Override
+    public Item addNewItem(Item item) {
+        itemRepository.save(item);
+        return getItem(item.getId());
```
Для чего отдельный запрос следом? Почему не подсмотреть айди сущности в ответе от репозитория?

Anarmard replied:
Правил эту строку, когда искал ошибку. Сейчас вернул как было, ошибок вроде нет.

alexache commented on src/main/java/ru/practicum/shareit/item/service/ItemServiceImpl.java:
```diff
@@ -2,54 +2,145 @@
+    @Override
+    public CommentResponseDto addComment(Comment comment) {
+
+        if (bookingRepository.findPastByBooker(comment.getAuthor(), LocalDateTime.now()).isEmpty()) {
```
Хорошо бы еще проверять, что автор комментария имел бронирования именно той же вещи, что и вещь в комментарии

Anarmard replied:
Да, конечно! не проверял, что именно комментируемую вещь он бронировал. Поправил, спасибо.

alexache commented on src/main/java/ru/practicum/shareit/request/model/ItemRequest.java:
```diff
@@ -1,17 +1,31 @@
 public class ItemRequest {
+
+    @Id
+    //@SequenceGenerator(name = "pk_sequence", schema = "public", sequenceName = "requests_id_seq", allocationSize = 1)
```
Закомментированный код лучше подчищать

Anarmard replied:
Да, извини, забыл убрать в этом разделе.

alexache commented on src/main/java/ru/practicum/shareit/user/service/UserServiceImpl.java:
```diff
@@ -29,11 +33,36 @@ public User saveUser(User user) {
     }
+
+    public boolean checkIsEmailUnique(User user) {
+        for (User u: getAllUsers()) {
```
Сложный способ. а отчего бы не сделать запрос к базе для поиска юзера по имейлу?

Anarmard replied:
Да, спасибо за совет! Поправил.

Anarmard committed:
v2.12
- исправил код после комментариев ревьюера (часть 1)

Anarmard committed:
v2.13
Исправил код после комментариев ревьюера (часть 2):
- UserServiceImpl: Оптимизация проверки уникальности Email;
- ItemServiceImpl: addComment. Проверять, что автор букировал именно ту вещь, которую комментирует
- ItemServiceImpl: getItemsBooking. Нельзя в цикле обращаться к БД.

Anarmard committed:
v2.13.1
Исправил код после комментариев ревьюера (часть 2):
- UserServiceImpl: Оптимизация проверки уникальности Email;
- ItemServiceImpl: addComment. Проверять, что автор букировал именно ту вещь, которую комментирует
- ItemServiceImpl: getItemsBooking. Нельзя в цикле обращаться к БД.

Anarmard committed:
v2.14
Исправил код после комментариев ревьюера (часть 3):
RequestParam (default value)

Anarmard commented:
Александр, спасибо за комментарии! Поправил почти везде. По оставшимся опциональным вопросам есть сомнения как лучше сделать.

Anarmard committed:
v2.15
Контроллеры облегчены, всё перенесено в сервисы

alexache commented on src/main/java/ru/practicum/shareit/booking/service/BookingServiceImpl.java:
```diff
@@ -20,20 +27,41 @@
+        userService.getUserDtoById(userId);
+
+        // получение Item по её идентификатору из БД
+        ItemBookingResponseDto currentItem = itemService.getItemBooking(bookingCreateRequestDto.getItemId(), userId);
```
На будущее так же стоит придерживаться такого подхода;
 - контроллеры взаимодействуют с сервисами
 - сервисы принимают/возвращают ДТО (это опционально, контроллер тоже может использовать маппер. но единообразно)
 - сервисы взаимодействуют с репозиториями (потому что сервисы возвращают ДТО, но самим сервисам удобнее работать с моделями)

Anarmard replied:
Александр, по поводу пункта 3 такой вопрос на примере getItemBooking: 
в этом методе не только обращение к репозитарию ItemRepository, но и много других проверок, циклов и т.д.... Не понимаю, как в данном случае сервису BookingServiceImpl взаимодействовать с ItemRepository, а не с ItemServiceImpl.. 
Или я не понял комментарий

Anarmard replied:
Я поправлю данный момент до merge

alexache commented on src/main/java/ru/practicum/shareit/booking/service/BookingServiceImpl.java:
```diff
@@ -20,20 +27,41 @@
+        Booking currentBooking = bookingMapper.toBooking(bookingCreateRequestDto, userId);
+
+        // перевод Item из DTO в модель + установка Item в текущий Booking
+        currentBooking.setItem(itemMapper.toItem(currentItem, currentItem.getOwner().getId()));
```
Вот здесь получается лишний вызов маппера - выше в сервисе вещей была сделана обратная конвертациия модели в дто, а здесь уже обратно, это лишняя работа. Если бы сервис бронирований работал с репозиторием вещей, такой проблемы бы не было

Anarmard replied:
см. мой предыдущий комментарий. Не понимаю как в данном случае обойтись без ItemSerivceImpl...

alexache commented on src/main/java/ru/practicum/shareit/item/dto/CommentCreateRequestDto.java:
```diff
@@ -21,7 +19,7 @@ public class CommentCreateRequestDto {
 
     private Item item;
```
Это осталось в виде модели, почему бы не изменить на ДТО?

Anarmard replied:
Да, извини. Вроде смотрел все дто по данному вопросу, но проглядел((

alexache commented on src/main/java/ru/practicum/shareit/item/service/ItemServiceImpl.java:
```diff
@@ -82,23 +86,101 @@ public ItemBookingResponseDto getItemBooking(Long itemId, Long userId) {
+
+            // 1. ищем lastbooking для текущего Item
+            Booking lastBooking = null;
+            for (Booking currentLastBooking : lastListBooking) {
```
Не очень оптимальный способ с проходом списка во вложенном цикле. Это квадратичная сложность, то есть довольно медленно с ростом размера списка. Как можно сделать иначе? Переложить сначала last/next booking в мапы единственным проходом по списку, это будет линейная сложность

Anarmard replied:
Понял в чем проблема, но не уверен, что понял, как надо правильно сделать) Посмотри, пожалуйста. Вроде нет теперь цикла в цикле. 
Можно наверное еще через стрим наверное передать из List в Map.

alexache commented on src/main/java/ru/practicum/shareit/user/service/UserServiceImpl.java:
```diff
@@ -14,25 +18,32 @@
+        return userMapper.toUserDto(getUserById(userId));
+    }
+
     public User getUserById(Long userId) {
```
Это получается приватный метод?

Anarmard replied:
Да, спасибо, поправил!

alexache commented on src/main/java/ru/practicum/shareit/user/service/UserServiceImpl.java:
```diff
@@ -56,13 +67,13 @@ public void deleteUser(Long userId) {
+            // не нашли юзера с таким email
+            return true;
+        } else {
+            User u = repository.findUserByEmail(user.getEmail());
```
Зачем второй раз одинаковый запрос к базе? Можно с первого раза сохранить результат?

Anarmard replied:
Точно)) что-то опять затупил...

Anarmard committed:
v2.16
Убрал цикл в цикле
ItemServiceImpl

Anarmard commented:
В следующей версии переделать:
Сервисы должны вызывать репозиторий, а не другой сервис, если необходимо получить данные.
